### PR TITLE
feat: Async jobs

### PR DESCRIPTION
**Pull request by ramedina86**
This PR adds capabilities for async jobs API, building on my previous PR that enabled `isSafe` for events.

Running a workflow is a 2 step process:
- A `POST` endpoint is used to create the job. This endpoint provides a job `id` and a job `token`.
- Using the `id` and `token` provided in the previous step, the user can poll a `GET` endpoint until the result becomes available. 

To accommodate for multi-node setups, persistence of job runs is provided via Redis, with an in-memory `JobVault` for debugging/local development purposes.

- The main area of interest is `serve.py`, which has the endpoints.
- `crypto.py` generates the necessary hashes.
- `app_runner.py` was modified to accommodate for a new message, `hashRequest`. This is used by the frontend to generate the full URLs when the button `Call via API` is clicked.

Changes to workflow blocks don't need to be reviewed. They're small changes to restrict the way workflows are executed (only `payload` can be configured now, as opposed to the whole execution environment). Also I added some changes to make JSON parsing more resilient.

New button in Settings (only shows for Workflows, deactivated if workflow key is empty):
Screenshot 2024-12-11 at 13 37 06

### CODE REVIEW

**Review by FabienArcellier — COMMENTED** [review]
Still in progress

File: src/writer/crypto.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+def get_hash(message: str):
+    base_hash = os.getenv("WRITER_BASE_HASH")
+    if not base_hash:
+        raise ValueError("Environment variable WRITER_BASE_HASH needs to be set up in" + \
+                            "order to enable operations which require hash generation, such as creating async jobs.")
+    assert HASH_SALT
+    combined = base_hash + HASH_SALT + message
+    return hashlib.sha256(combined.encode()).hexdigest()
+
+def verify_hash_in_request(message: str, request: Request):
```
Comment (FabienArcellier) [inline src/writer/crypto.py:18]: I would rename the function `server:_verify_hash_in_request(message, request)` into `verify_message_authorization_signature(message, request)`

**Comment by FabienArcellier** [comment]
## prepare a documentation section to protect the endpoint of the job api

I would recommend detailing how to protect the `api/job` endpoint using Fastapi middleware to implement an API token. I am not sure the classical approach of using dependency won't work in our situation.

* https://fastapi.tiangolo.com/tutorial/middleware/?h=middle

**Review by FabienArcellier — APPROVED** [review]
In the end, I find the pattern consistent. I can see how to extend it to either using event handlers with perhaps a specific decorator `@writer.workflow(name="wf1")` that will generate a workflow as we generate backend ui.

File: src/writer/crypto.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+
```
[the remark this answers is not in the document]
Comment (ramedina86) [reply src/writer/crypto.py:1]: Yes, that was my idea. I've renamed it though, as I think `WRITER_SECRET_KEY` is more descriptive and highlights that it's indeed a very sensitive thing.

File: src/writer/serve.py [hunk 3]
```diff
@@ -139,7 +189,7 @@ async def lifespan(asgi_app: FastAPI):
     """
     app.state.writer_app = True
     app.state.app_runner = app_runner
-    app.state.job_vault = JobVault()
+    app.state.job_vault = JobVault.create_vault()
```
[the remark this answers is not in the document]
Comment (ramedina86) [reply src/writer/serve.py]: Nice one, implemented.

File: src/writer/crypto.py (hunk 1 above)
Comment (ramedina86) [reply src/writer/crypto.py]: It's clearer yes, implemented.

**Comment by ramedina86** [comment]
Thanks a lot for the great insights @FabienArcellier 🙏🏻

I've already implemented most of it, as commented above.

> enable the workflows API through the command line or get_asgi_app
Makes sense for this feature to be disabled by default. My idea was that if `WRITER_BASE_HASH` (now known as `WRITER_SECRET_KEY`) was present, the jobs API feature was considered activated. However, it's better to make it explicit. I'll implement. 

> prepare a documentation section to protect the endpoint of the job api
We'll do this when we document Async Jobs API in general (for now it'll remain undocumented).

**Pull request merged**